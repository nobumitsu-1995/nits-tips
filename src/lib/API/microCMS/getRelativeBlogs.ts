import type { MicroCMS } from '@/types/microCMS'
import { getBlogsData } from './getBlogs'

type Blog = MicroCMS['blog']

/** 関連記事の優先順位
 * タイトルが類似 > タグが同じ > カテゴリが同じ
 */
export const getRelativeBlogs = async (blog: Blog) => {
  const titleRelatives =
    (await getBlogsData({
      filters: `title[contains]${blog.title}[and]id[not_equals]${blog.id}`,
      limit: 4,
    })) ?? []

  const tagRelatives =
    (await getBlogsData({
      filters: `tags[contains]${blog.tags[0].id}[and]id[not_equals]${blog.id}`,
      limit: 4,
    })) ?? []

  const categoryRelatives =
    (await getBlogsData({
      filters: `category[equals]${blog.category.id}[and]id[not_equals]${blog.id}`,
      limit: 4,
    })) ?? []

  // 重複した記事を取得する可能性があるので以下の処理でユニークな値を持つ配列にする。
  const map = new Map<string, Blog>()
  const arr = [...titleRelatives, ...tagRelatives, ...categoryRelatives]
  arr.forEach((obj) => {
    map.set(obj.id, obj)
  })

  // 関連記事の表示は4件までなので丸め込む
  return Array.from(map.values()).slice(0, 4)
}
