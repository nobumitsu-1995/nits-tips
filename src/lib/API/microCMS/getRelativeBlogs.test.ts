import { blog } from '@/testUtils/fixtures/blogsData'
import { getRelativeBlogs } from './getRelativeBlogs'

const mockedJson = vitest.fn().mockResolvedValue({})
const mockedFetch = vitest.fn()

beforeEach(() => {
  const mockResponse = {
    ok: true,
    status: 200,
    json: mockedJson,
  }
  global.fetch = mockedFetch.mockResolvedValue(mockResponse)
})

afterEach(() => {
  vitest.resetAllMocks()
})

test('getRelativeBlogsで正常に値が取得できる', async () => {
  await getRelativeBlogs(blog())
  expect(await mockedFetch()).toMatchSnapshot()
  expect(await mockedFetch()).toMatchSnapshot()
  expect(await mockedFetch()).toMatchSnapshot()
  expect(mockedJson).toHaveBeenCalledWith()
})
