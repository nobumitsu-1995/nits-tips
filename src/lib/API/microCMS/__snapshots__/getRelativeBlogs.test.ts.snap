// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getRelativeBlogsで正常に値が取得できる 1`] = `
{
  "json": [MockFunction spy] {
    "calls": [
      [],
      [],
      [],
    ],
    "results": [
      {
        "type": "return",
        "value": Promise {},
      },
      {
        "type": "return",
        "value": Promise {},
      },
      {
        "type": "return",
        "value": Promise {},
      },
    ],
  },
  "ok": true,
  "status": 200,
}
`;

exports[`getRelativeBlogsで正常に値が取得できる 2`] = `
{
  "json": [MockFunction spy] {
    "calls": [
      [],
      [],
      [],
    ],
    "results": [
      {
        "type": "return",
        "value": Promise {},
      },
      {
        "type": "return",
        "value": Promise {},
      },
      {
        "type": "return",
        "value": Promise {},
      },
    ],
  },
  "ok": true,
  "status": 200,
}
`;

exports[`getRelativeBlogsで正常に値が取得できる 3`] = `
{
  "json": [MockFunction spy] {
    "calls": [
      [],
      [],
      [],
    ],
    "results": [
      {
        "type": "return",
        "value": Promise {},
      },
      {
        "type": "return",
        "value": Promise {},
      },
      {
        "type": "return",
        "value": Promise {},
      },
    ],
  },
  "ok": true,
  "status": 200,
}
`;
